table/transaction.go

```diff
@@ -0,0 +1,346 @@
+}
+
+func (sc *schemaCompatVisitor) Primitive(p iceberg.PrimitiveType) bool {
+	return true
+}
+
+type snapshotUpdate struct {
+	txn           *Transaction
+	io            io.WriteFileIO
+	snapshotProps iceberg.Properties
+}
+
+func (s snapshotUpdate) fastAppend() *snapshotProducer {
+	return newFastAppendFilesProducer(OpAppend, s.txn, s.io, nil, s.snapshotProps)
+}
+
+func (s snapshotUpdate) mergeAppend() *snapshotProducer {
+	return newMergeAppendFilesProducer(OpAppend, s.txn, s.io, nil, s.snapshotProps)
+}
+
+type Transaction struct {
+	tbl  *Table
+	meta *MetadataBuilder
+
+	reqs []Requirement
+
+	mx        sync.Mutex
+	committed bool
+}
+
+func (t *Transaction) apply(updates []Update, reqs []Requirement) error {
+	t.mx.Lock()
+	defer t.mx.Unlock()
+
+	if t.committed {
+		return errors.New("transaction has already been committed")
+	}
+
+	current, err := t.meta.Build()
+	if err != nil {
+		return err
+	}
+
+	for _, r := range reqs {
+		if err := r.Validate(current); err != nil {
+			return err
+		}
+	}
+
+	existing := map[string]struct{}{}
+	for _, r := range t.reqs {
+		existing[r.GetType()] = struct{}{}
+	}
+
+	for _, r := range reqs {
+		if _, ok := existing[r.GetType()]; !ok {
+			t.reqs = append(t.reqs, r)
+		}
+	}
+
+	prevUpdates, prevLastUpdated := len(t.meta.updates), t.meta.lastUpdatedMS
+	for _, u := range updates {
+		if err := u.Apply(t.meta); err != nil {
+			return err
+		}
+	}
+
+	if prevUpdates < len(t.meta.updates) {
+		if prevLastUpdated == t.meta.lastUpdatedMS {
+			t.meta.lastUpdatedMS = time.Now().UnixMilli()
+		}
+	}
+
+	return nil
+}
+
+func (t *Transaction) updateSnapshot(props iceberg.Properties) snapshotUpdate {
+	return snapshotUpdate{
+		txn:           t,
+		io:            t.tbl.fs.(io.WriteFileIO),
+		snapshotProps: props,
+	}
+}
+
+func (t *Transaction) appendSnapshotProducer(props iceberg.Properties) *snapshotProducer {
+	manifestMerge := t.meta.props.GetBool(ManifestMergeEnabledKey, ManifestMergeEnabledDefault)
+
+	updateSnapshot := t.updateSnapshot(props)
+	if manifestMerge {
+		return updateSnapshot.mergeAppend()
+	}
+
+	return updateSnapshot.fastAppend()
+}
+
+func (t *Transaction) SetProperties(props iceberg.Properties) error {
+	if len(props) > 0 {
+		return t.apply([]Update{NewSetPropertiesUpdate(props)}, nil)
+	}
+
+	return nil
+}
+
+func (t *Transaction) Append(rdr array.RecordReader, snapshotProps iceberg.Properties) error {
+	if err := checkArrowSchemaCompat(t.meta.CurrentSchema(), rdr.Schema(), false); err != nil {
+		return err
+	}
+
+	t.appendSnapshotProducer(snapshotProps)
+
+	return nil
+}
+
+func (t *Transaction) AddFiles(files []string, snapshotProps iceberg.Properties, checkDuplicates bool) error {
+	set := make(map[string]string)
+	for _, f := range files {
+		set[f] = f
+	}
+
+	if len(set) != len(files) {
+		return errors.New("file paths must be unique for AppendDataFiles")
+	}
+
+	if checkDuplicates {
+		// TODO: implement
```

Comment:
By exposing this, an user might assume that this is already implemented. Maybe we should raise here?

Reply:
In a subsequent PR I implement this. But you're right, I'll have it return a not implemented error here for now.